shards/shards.go

```diff
@@ -367,51 +428,34 @@ func (ss *shardedSearcher) Search(ctx context.Context, q query.Q, opts *zoekt.Se
 		feeder <- s
 	}
 	close(feeder)
+	g := errgroup.Group{}
 	for i := 0; i < runtime.GOMAXPROCS(0); i++ {
-		go func() {
+		g.Go(func() error {
 			for s := range feeder {
-				searchOneShard(childCtx, s, q, opts, all)
-			}
-		}()
-	}
-
-	for range shards {
-		r := <-all
-		if r.err != nil {
-			return nil, r.err
-		}
-		aggregate.Files = append(aggregate.Files, r.sr.Files...)
-		aggregate.Stats.Add(r.sr.Stats)
-
-		if len(r.sr.Files) > 0 {
-			for k, v := range r.sr.RepoURLs {
-				aggregate.RepoURLs[k] = v
-			}
-			for k, v := range r.sr.LineFragments {
-				aggregate.LineFragments[k] = v
+				err := searchOneShard(childCtx, s, q, opts, stream.SenderFunc(func(sr *zoekt.SearchResult) {
+					if sr != nil {
+						metricSearchContentBytesLoadedTotal.Add(float64(sr.Stats.ContentBytesLoaded))
+						metricSearchIndexBytesLoadedTotal.Add(float64(sr.Stats.IndexBytesLoaded))
+						metricSearchCrashesTotal.Add(float64(sr.Stats.Crashes))
+						metricSearchFileCountTotal.Add(float64(sr.Stats.FileCount))
+						metricSearchShardFilesConsideredTotal.Add(float64(sr.Stats.ShardFilesConsidered))
+						metricSearchFilesConsideredTotal.Add(float64(sr.Stats.FilesConsidered))
+						metricSearchFilesLoadedTotal.Add(float64(sr.Stats.FilesLoaded))
+						metricSearchFilesSkippedTotal.Add(float64(sr.Stats.FilesSkipped))
+						metricSearchShardsSkippedTotal.Add(float64(sr.Stats.ShardsSkipped))
+						metricSearchMatchCountTotal.Add(float64(sr.Stats.MatchCount))
+						metricSearchNgramMatchesTotal.Add(float64(sr.Stats.NgramMatches))
```

Comment:
given we are now doing this per shard, I wonder if the overhead of recording all these metrics will add up.

Reply:
Yeah, I was thinking the same. However, if we want to stream out to Sourcegraph without any aggregation we have to do it either per event or drop it completely. We could remove it for streaming IE move this block from `streamingSearch` to `Search`.

Reply:
I think lets leave it as is. We can aggregate in a stats object then log it finally if we notice a locking overhead (I think each call acquires a mutex?)